File: Bronze1-1/06-24-am/unit09/group-sum/groupsum.cpp
/*
 * Group sum
 */
#include <bits/stdc++.h>
using namespace std;

//Report whether to find a subsequence of A (from "start" to the end, 
// not necessarily continuous) that adds up to sum.
bool solve(int A[], int start, int sum, int len) {
	if(sum==0)
		return true;
	
	if(start==len-1)
		return A[start]==sum;
	
	return solve(A, start+1, sum, len) || solve(A, start+1, sum-A[start], len);
}

int main() {
	int Q; cin >> Q; //# of queries
	while(Q--) {
		int N, T;
		cin >> N >> T;
		
		int A[N];
		for(int j=0; j<N; j++)
			cin >> A[j];

		bool res = solve(A, 0, T, N);
		cout << (res ? "YES" : "NO") << '\n';
	}
}

File: Bronze1-1/06-24-am/unit09/group-sum/groupsum.java
/*
 * Group sum
 */
import java.io.*;
import java.util.*;
public class groupsum {
	public static void main(String[] args) throws IOException {
		Scanner in = new Scanner(System.in);
		int Q = in.nextInt(); //# of queries
		while(Q-->0) {
			int N = in.nextInt();
			int T = in.nextInt();
			
			int[] A = new int[N];
			for(int j=0; j<N; j++)
				A[j] = in.nextInt();
	
			boolean res = sum(A, 0, T);
			System.out.println(res ? "YES" : "NO");
		}
		System.out.println();
		in.close();
	}

	//Report whether to find a subsequence of A (from "start" to the end, 
	// not necessarily continuous) that adds up to sum.
	public static boolean sum(int[] A, int start, int sum) {
		if(sum==0)
			return true;
		
		if(start==A.length-1)
			return A[start]==sum;
		
		return sum(A, start+1, sum) || sum(A, start+1, sum-A[start]);
	}
}
